[LT] Merge permutation_util into master

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67929
* __->__ #67766

Test Plan: `build/bin/test_lazy`

Differential Revision: [D32147676](https://our.internmc.facebook.com/intern/diff/D32147676)

desertfire committed:
[LT] Merge permutation_util into master

Test Plan: `build/bin/test_lazy`

[ghstack-poisoned]

desertfire commented on tools/build_variables.bzl:
```diff
@@ -360,6 +360,7 @@ lazy_tensor_core_sources = [
     "torch/csrc/lazy/core/ir.cpp",
     "torch/csrc/lazy/core/ir_metadata.cpp",
     "torch/csrc/lazy/core/ir_util.cpp",
+    "torch/csrc/lazy/tensor/permutation_util.cpp",
```
@wconstab I put the new file in a new subdirectory. Want to hear what's your plan in organizing files here. Do you prefer putting everything under `core`? Or is `core` just for IR nodes?

wconstab replied:
I don't have a grand plan;  I think originally I did a little planning and came up with /core and /backend as two folders.  Currently I'm not sure I like it anymore; In general I think I'd lean towards putting everything into one folder first, then reorganizing it once we have enough files to really justify a separate folder.  

What do you think about deleting the 'core' folder entirely and just putting everything under torch/lazy for now?  Except for a separate torch/lazy/ts folder for the backend impl?
torch/lazy (most stuff here)
torch/lazy/ts_backend/ (backend impl)

Or, do you like core and maybe
torch/lazy/core/ (everything else)
torch/lazy/backend/ (backend interface)
torch/lazy/ts_backend (backend implementation)

desertfire replied:
I like keeping torch/lazy/core/ as a subdirectory. Let me move the new files into the same directory.

desertfire committed:
Update on "[LT] Merge permutation_util into master"


Test Plan: `build/bin/test_lazy`

[ghstack-poisoned]

wconstab reviewed (approved):
This looks OK to me.  I didn't check how it is used in our codebase, is it used for View classes?  Assuming we need such a util and there isn't one in torch/c10, it seems fine as-is.

desertfire commented:
> This looks OK to me. I didn't check how it is used in our codebase, is it used for View classes? Assuming we need such a util and there isn't one in torch/c10, it seems fine as-is.

Yes, it is mostly used by the View class and view-like op nodes.

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32147676).

desertfire committed:
Update on "[LT] Merge permutation_util into master"


Test Plan: `build/bin/test_lazy`

Differential Revision: [D32147676](https://our.internmc.facebook.com/intern/diff/D32147676)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32147676).

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32147676).

desertfire committed:
Update on "[LT] Merge permutation_util into master"


Test Plan: `build/bin/test_lazy`

Differential Revision: [D32147676](https://our.internmc.facebook.com/intern/diff/D32147676)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32147676).

desertfire committed:
Update on "[LT] Merge permutation_util into master"


Test Plan: `build/bin/test_lazy`

Differential Revision: [D32147676](https://our.internmc.facebook.com/intern/diff/D32147676)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32147676).
